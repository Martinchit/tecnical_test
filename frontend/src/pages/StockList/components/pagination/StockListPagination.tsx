import React from 'react';
import { Pagination } from 'react-bootstrap';

interface StockListPaginationProps {
  currentPage: number;
  totalPages: number;
  onPaginationButtonClickHandler: (page: number) => void;
}

export const StockListPagination: React.FC<StockListPaginationProps> = ({
  currentPage,
  totalPages,
  onPaginationButtonClickHandler,
}) => {
  const ref = currentPage + 1 >= totalPages ? totalPages - currentPage + 3 : 5;
  const ellipsisMin = currentPage - 5 < 1 ? 1 : currentPage - 5;
  const ellipsisMax = currentPage + 5 > totalPages ? totalPages : currentPage + 5;
  return (
    <Pagination>
      <Pagination.First onClick={() => onPaginationButtonClickHandler(1)} />
      {currentPage > 3 ? (
        <>
          <Pagination.Item onClick={() => onPaginationButtonClickHandler(1)}>
            {1}
          </Pagination.Item>
          <Pagination.Ellipsis
            onClick={() => onPaginationButtonClickHandler(ellipsisMin)}
          />
        </>
      ) : null}
      {totalPages > 4
        ? [...Array(ref)].map((_, i) => {
            const p = currentPage < 3 ? 1 + i : currentPage + (-2 + i);
            return (
              <Pagination.Item
                onClick={() => onPaginationButtonClickHandler(p)}
                key={p}
                active={currentPage === p}
              >
                {p}
              </Pagination.Item>
            );
          })
        : [...Array(totalPages)].map((_, i) => {
            const p = currentPage < 3 ? 1 + i : currentPage + (-2 + i);
            return (
              <Pagination.Item
                onClick={() => onPaginationButtonClickHandler(p)}
                key={p}
                active={currentPage === p}
              >
                {p}
              </Pagination.Item>
            );
          })}
      {currentPage < totalPages - 2 && totalPages > 5 ? (
        <>
          <Pagination.Ellipsis
            onClick={() => onPaginationButtonClickHandler(ellipsisMax)}
          />
          <Pagination.Item
            onClick={() => onPaginationButtonClickHandler(totalPages)}
          >
            {totalPages}
          </Pagination.Item>
        </>
      ) : null}
      <Pagination.Last onClick={() => onPaginationButtonClickHandler(totalPages)} />
    </Pagination>
  );
};
