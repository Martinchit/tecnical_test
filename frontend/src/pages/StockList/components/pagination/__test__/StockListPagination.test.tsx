import React from 'react';
import { shallow } from 'enzyme';
import { Pagination } from 'react-bootstrap';

import { StockListPagination } from '../StockListPagination';

describe('Alert', () => {
  const props = {
    currentPage: 1,
    totalPages: 10,
    onPaginationButtonClickHandler: jest.fn(),
  };

  const restoreMock = () => {
    props.onPaginationButtonClickHandler.mockRestore();
  };

  it('renders correctly', () => {
    const wrapper = shallow(<StockListPagination {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('triggers onPaginationButtonClickHandler', () => {
    const wrapper = shallow(<StockListPagination {...props} />);
    wrapper.find(Pagination.First).first().simulate('click');
    expect(props.onPaginationButtonClickHandler).toHaveBeenCalledTimes(1);
    expect(props.onPaginationButtonClickHandler).toHaveBeenCalledWith(1);
    wrapper.find(Pagination.Last).first().simulate('click');
    expect(props.onPaginationButtonClickHandler).toHaveBeenCalledTimes(2);
    expect(props.onPaginationButtonClickHandler).toHaveBeenCalledWith(
      props.totalPages
    );
    restoreMock();
  });

  it('renders Pagination.Ellipsis', () => {
    const currentPage = 4;
    const wrapper = shallow(
      <StockListPagination
        totalPages={props.totalPages}
        onPaginationButtonClickHandler={props.onPaginationButtonClickHandler}
        currentPage={currentPage}
      />
    );
    expect(wrapper.find(Pagination.Ellipsis).length).toBe(2);
    wrapper.find(Pagination.Ellipsis).first().simulate('click');
    expect(props.onPaginationButtonClickHandler).toHaveBeenCalledWith(1);
    restoreMock();
    wrapper.find(Pagination.Item).first().simulate('click');
    expect(props.onPaginationButtonClickHandler).toHaveBeenCalledWith(1);
    restoreMock();
    wrapper.find(Pagination.Ellipsis).last().simulate('click');
    expect(props.onPaginationButtonClickHandler).toHaveBeenCalledWith(4 + 5);
    restoreMock();
    wrapper.find(Pagination.Item).last().simulate('click');
    expect(props.onPaginationButtonClickHandler).toHaveBeenCalledWith(
      props.totalPages
    );
    restoreMock();
  });

  it('renders correctly with less than 5 pages', () => {
    const totalPages = 3;
    const wrapper = shallow(
      <StockListPagination
        totalPages={totalPages}
        onPaginationButtonClickHandler={props.onPaginationButtonClickHandler}
        currentPage={props.currentPage}
      />
    );
    expect(wrapper.find(Pagination.Item).length).toBe(totalPages);
    wrapper.find(Pagination.Item).first().simulate('click');
    expect(props.onPaginationButtonClickHandler).toHaveBeenCalledWith(1);
    restoreMock();
    wrapper.find(Pagination.Item).at(1).simulate('click');
    expect(props.onPaginationButtonClickHandler).toHaveBeenCalledWith(2);
    restoreMock();
    wrapper.find(Pagination.Item).at(2).simulate('click');
    expect(props.onPaginationButtonClickHandler).toHaveBeenCalledWith(3);
    restoreMock();
  });

  it('renders correctly with more than 5 pages', () => {
    const totalPages = 10;
    const currentPage = 5;
    const wrapper = shallow(
      <StockListPagination
        totalPages={totalPages}
        onPaginationButtonClickHandler={props.onPaginationButtonClickHandler}
        currentPage={currentPage}
      />
    );
    expect(wrapper.find(Pagination.Item).length).toBe(7);
    wrapper.find(Pagination.Item).at(1).simulate('click');
    expect(props.onPaginationButtonClickHandler).toHaveBeenCalledWith(3);
    restoreMock();
    wrapper.find(Pagination.Item).at(2).simulate('click');
    expect(props.onPaginationButtonClickHandler).toHaveBeenCalledWith(4);
    restoreMock();
    wrapper.find(Pagination.Ellipsis).first().simulate('click');
    expect(props.onPaginationButtonClickHandler).toHaveBeenCalledWith(1);
    restoreMock();
    wrapper.find(Pagination.Ellipsis).last().simulate('click');
    expect(props.onPaginationButtonClickHandler).toHaveBeenCalledWith(10);
    restoreMock();
  });
});
