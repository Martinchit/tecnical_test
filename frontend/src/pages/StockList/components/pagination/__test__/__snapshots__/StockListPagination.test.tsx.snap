// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert renders correctly 1`] = `
<ForwardRef>
  <First
    onClick={[Function]}
  />
  <PageItem
    active={true}
    activeLabel="(current)"
    disabled={false}
    key="1"
    onClick={[Function]}
  >
    1
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="2"
    onClick={[Function]}
  >
    2
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="3"
    onClick={[Function]}
  >
    3
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="4"
    onClick={[Function]}
  >
    4
  </PageItem>
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    key="5"
    onClick={[Function]}
  >
    5
  </PageItem>
  <Ellipsis
    onClick={[Function]}
  />
  <PageItem
    active={false}
    activeLabel="(current)"
    disabled={false}
    onClick={[Function]}
  >
    10
  </PageItem>
  <Last
    onClick={[Function]}
  />
</ForwardRef>
`;
